File: smoke_tests/kotlin/mass_assignment.kt
// Mass Assignment vulnerabilities in Kotlin
package com.example.security

import com.fasterxml.jackson.databind.ObjectMapper

class MassAssignmentVulnerabilities {

    private val objectMapper = ObjectMapper()

    // Test 1: Direct map assignment
    fun updateUser(userId: Int, params: Map<String, Any>) {
        val user = getUser(userId)
        // VULNERABLE: All params assigned
        params.forEach { (key, value) ->
            setUserProperty(user, key, value)
        }
        saveUser(user)
    }

    // Test 2: Jackson deserialization
    fun createUserFromJson(json: String): User {
        // VULNERABLE: Decodes all fields including isAdmin
        return objectMapper.readValue(json, User::class.java)
    }

    // Test 3: Data class copy with spread
    fun patchUser(userId: Int, updates: Map<String, Any>): User {
        val user = getUser(userId)
        // VULNERABLE: Copying with untrusted updates
        return user.copy(
            name = updates["name"] as? String ?: user.name,
            email = updates["email"] as? String ?: user.email,
            isAdmin = updates["isAdmin"] as? Boolean ?: user.isAdmin, // VULNERABLE
            role = updates["role"] as? String ?: user.role // VULNERABLE
        )
    }

    // Test 4: Reflection-based assignment
    fun updateObject(obj: Any, params: Map<String, Any>) {
        // VULNERABLE: Setting arbitrary properties
        params.forEach { (key, value) ->
            val field = obj.javaClass.getDeclaredField(key)
            field.isAccessible = true
            field.set(obj, value)
        }
    }

    // Test 5: Form data binding
    fun handleFormSubmission(formData: Map<String, String>): Profile {
        // VULNERABLE: All form fields accepted
        return Profile(
            name = formData["name"] ?: "",
            bio = formData["bio"] ?: "",
            isVerified = formData["isVerified"]?.toBoolean() ?: false, // VULNERABLE
            permissions = formData["permissions"] ?: "" // VULNERABLE
        )
    }

    // Test 6: API request body
    fun handleApiRequest(body: Map<String, Any>): Map<String, Any> {
        // VULNERABLE: Copying all body params
        val userData = body["user"] as? Map<String, Any>
        if (userData != null) {
            createUser(userData)
        }
        return mapOf("status" to "created")
    }

    // Test 7: Apply extension function abuse
    fun updateSettings(settings: Settings, params: Map<String, Any>): Settings {
        // VULNERABLE: All params applied
        return settings.apply {
            params["theme"]?.let { theme = it as String }
            params["language"]?.let { language = it as String }
            params["adminMode"]?.let { adminMode = it as Boolean } // VULNERABLE
        }
    }

    // Test 8: Constructor with defaults
    fun createUserFromParams(params: Map<String, Any>): User {
        // VULNERABLE: All params used in construction
        return User(
            id = params["id"] as? Int ?: 0,
            name = params["name"] as? String ?: "",
            email = params["email"] as? String ?: "",
            isAdmin = params["isAdmin"] as? Boolean ?: false,
            role = params["role"] as? String ?: "user"
        )
    }

    // Test 9: GSON deserialization
    fun deserializeWithGson(json: String): User {
        // VULNERABLE: All fields decoded
        return com.google.gson.Gson().fromJson(json, User::class.java)
    }

    // Test 10: Merge maps
    fun mergeUserData(userId: Int, newData: Map<String, Any>): Map<String, Any> {
        val existingData = getUserData(userId).toMutableMap()
        // VULNERABLE: Merging untrusted data
        existingData.putAll(newData)
        return existingData
    }

    // Test 11: Kotlin serialization
    fun deserializeKotlinx(json: String): User {
        // VULNERABLE: All fields deserialized
        return kotlinx.serialization.json.Json.decodeFromString(json)
    }

    // Test 12: Builder pattern with untrusted input
    fun buildUser(params: Map<String, Any>): User {
        // VULNERABLE: Builder accepts all params
        return UserBuilder()
            .id(params["id"] as? Int ?: 0)
            .name(params["name"] as? String ?: "")
            .isAdmin(params["isAdmin"] as? Boolean ?: false) // VULNERABLE
            .build()
    }

    private fun getUser(userId: Int): User = User(userId, "", "", false, "user")
    private fun setUserProperty(user: User, key: String, value: Any) {}
    private fun saveUser(user: User) {}
    private fun createUser(data: Map<String, Any>) {}
    private fun getUserData(userId: Int): Map<String, Any> = emptyMap()
}

data class User(
    val id: Int,
    var name: String,
    var email: String,
    var isAdmin: Boolean,
    var role: String
)

data class Profile(
    val name: String,
    val bio: String,
    val isVerified: Boolean,
    val permissions: String
)

data class Settings(
    var theme: String = "light",
    var language: String = "en",
    var adminMode: Boolean = false
)

class UserBuilder {
    private var id: Int = 0
    private var name: String = ""
    private var isAdmin: Boolean = false

    fun id(id: Int) = apply { this.id = id }
    fun name(name: String) = apply { this.name = name }
    fun isAdmin(isAdmin: Boolean) = apply { this.isAdmin = isAdmin }
    fun build() = User(id, name, "", isAdmin, "user")
}

File: smoke_tests/scala/mass_assignment.scala
// Mass Assignment vulnerabilities in Scala
package com.example.security

import scala.collection.mutable

class MassAssignmentVulnerabilities {

  // Test 1: Direct map assignment
  def updateUser(userId: Int, params: Map[String, Any]): Unit = {
    val user = getUser(userId)
    // VULNERABLE: All params assigned
    params.foreach { case (key, value) =>
      setUserProperty(user, key, value)
    }
    saveUser(user)
  }

  // Test 2: Case class copy with spread
  def patchUser(userId: Int, updates: Map[String, Any]): User = {
    val user = getUser(userId)
    // VULNERABLE: Copying with untrusted updates
    user.copy(
      name = updates.getOrElse("name", user.name).asInstanceOf[String],
      email = updates.getOrElse("email", user.email).asInstanceOf[String],
      isAdmin = updates.getOrElse("isAdmin", user.isAdmin).asInstanceOf[Boolean], // VULNERABLE
      role = updates.getOrElse("role", user.role).asInstanceOf[String] // VULNERABLE
    )
  }

  // Test 3: Reflection-based assignment
  def updateObject(obj: AnyRef, params: Map[String, Any]): Unit = {
    // VULNERABLE: Setting arbitrary properties
    params.foreach { case (key, value) =>
      val field = obj.getClass.getDeclaredField(key)
      field.setAccessible(true)
      field.set(obj, value)
    }
  }

  // Test 4: Play JSON binding (conceptual)
  def handleJsonRequest(json: String): User = {
    // VULNERABLE: All fields from JSON
    import play.api.libs.json._
    implicit val userReads: Reads[User] = Json.reads[User]
    Json.parse(json).as[User]
  }

  // Test 5: Form data binding
  def handleFormSubmission(formData: Map[String, String]): Profile = {
    // VULNERABLE: All form fields accepted
    Profile(
      name = formData.getOrElse("name", ""),
      bio = formData.getOrElse("bio", ""),
      isVerified = formData.get("isVerified").exists(_.toBoolean), // VULNERABLE
      permissions = formData.getOrElse("permissions", "") // VULNERABLE
    )
  }

  // Test 6: Circe JSON decode
  def decodeWithCirce(json: String): User = {
    // VULNERABLE: Decoding all fields
    import io.circe.parser._
    import io.circe.generic.auto._
    decode[User](json).getOrElse(throw new Exception("Parse failed"))
  }

  // Test 7: Merge maps
  def mergeUserData(userId: Int, newData: Map[String, Any]): Map[String, Any] = {
    val existingData = getUserData(userId)
    // VULNERABLE: Merging untrusted data
    existingData ++ newData
  }

  // Test 8: Spray JSON format
  def sprayJsonFormat(json: String): User = {
    // VULNERABLE: All fields deserialized
    import spray.json._
    import DefaultJsonProtocol._
    implicit val userFormat: RootJsonFormat[User] = jsonFormat4(User.apply)
    json.parseJson.convertTo[User]
  }

  // Test 9: Argonaut decode
  def argonautDecode(json: String): User = {
    // VULNERABLE: Full object decode
    import argonaut._, Argonaut._
    json.decodeOption[User].get
  }

  // Test 10: Builder pattern abuse
  def buildFromParams(params: Map[String, Any]): User = {
    // VULNERABLE: Builder accepts all params
    UserBuilder()
      .id(params.getOrElse("id", 0).asInstanceOf[Int])
      .name(params.getOrElse("name", "").asInstanceOf[String])
      .isAdmin(params.getOrElse("isAdmin", false).asInstanceOf[Boolean]) // VULNERABLE
      .build()
  }

  // Test 11: Mutable object update
  def updateMutableUser(user: MutableUser, params: Map[String, Any]): Unit = {
    // VULNERABLE: All params used
    params.get("name").foreach(v => user.name = v.asInstanceOf[String])
    params.get("isAdmin").foreach(v => user.isAdmin = v.asInstanceOf[Boolean])
    params.get("role").foreach(v => user.role = v.asInstanceOf[String])
  }

  // Test 12: Slick entity update (conceptual)
  def slickUpdate(userId: Int, values: Map[String, Any]): Unit = {
    // VULNERABLE: All columns from map
    val updates = values.map { case (k, v) => s"$k = $v" }.mkString(", ")
    s"UPDATE users SET $updates WHERE id = $userId"
  }

  private def getUser(userId: Int): User = User(userId, "", "", false, "user")
  private def setUserProperty(user: User, key: String, value: Any): Unit = ()
  private def saveUser(user: User): Unit = ()
  private def getUserData(userId: Int): Map[String, Any] = Map.empty
}

case class User(id: Int, name: String, email: String, isAdmin: Boolean, role: String)

case class Profile(name: String, bio: String, isVerified: Boolean, permissions: String)

class MutableUser {
  var name: String = ""
  var isAdmin: Boolean = false
  var role: String = "user"
}

case class UserBuilder(
  id: Int = 0,
  name: String = "",
  isAdmin: Boolean = false
) {
  def id(v: Int): UserBuilder = copy(id = v)
  def name(v: String): UserBuilder = copy(name = v)
  def isAdmin(v: Boolean): UserBuilder = copy(isAdmin = v)
  def build(): User = User(id, name, "", isAdmin, "user")
}
